src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java

```diff
@@ -319,36 +327,55 @@ private CoverageProvider getCoverageProvider(
+        copyCoverageToJenkinsMaster(build, workspace, listener);
+        CoverageProvider coverageProvider = new XmlCoverageProvider(getCoverageReports(build), includeFiles);
+        coverageProvider.computeCoverageIfNeeded();
+        cleanupCoverageFilesOnJenkinsMaster(build);
 
-        CoverageProvider coberturaCoverage = InstanceProvider.getCoberturaCoverageProvider(build,
-                workspace, includeFiles, coverageReportPattern, logger);
-        if (coberturaCoverage != null) {
-            coverageProviders.add(coberturaCoverage);
-        }
-
-        CoverageProvider jacocoCoverage = InstanceProvider.getJacocoCoverageProvider(build,
-                workspace, includeFiles, coverageReportPattern, logger);
-        if (jacocoCoverage != null) {
-            coverageProviders.add(jacocoCoverage);
+        if (coverageProvider.hasCoverage()) {
+            return coverageProvider;
+        } else {
+            Logger logger = new Logger(listener.getLogger());
+            logger.info(UBERALLS_TAG, "No coverage results found");
+            return null;
         }
+    }
 
-        for (Iterator<CoverageProvider> i = coverageProviders.iterator(); i.hasNext(); ) {
-            CoverageProvider coverageProvider = i.next();
-
-            if (!coverageProvider.hasCoverage()) {
-                i.remove();
+    private void copyCoverageToJenkinsMaster(Run<?, ?> build, FilePath workspace, TaskListener listener) {
+        Logger logger = new Logger(listener.getLogger());
+        final FilePath moduleRoot = workspace;
+        final File buildCoberturaDir = build.getRootDir();
+        FilePath buildTarget = new FilePath(buildCoberturaDir);
+
+        if (moduleRoot != null) {
+            try {
+                int i = 0;
+                for (FilePath report : moduleRoot.list(coverageReportPattern)) {
+                    final FilePath targetPath = new FilePath(buildTarget, "coverage" + (i == 0 ? "" : i) + ".xml");
+                    report.copyTo(targetPath);
+                    i++;
+                }
+            } catch (InterruptedException | IOException e) {
+                e.printStackTrace();
+                logger.warn(COVERAGE_TAG, "Unable to copy coverage to " + buildTarget);
             }
         }
-        if (!coverageProviders.isEmpty()) {
-            CoverageProvider provider = coverageProviders.get(0);
-            logger.info(UBERALLS_TAG, "Selected Coverage Provider: " + provider);
-            return provider;
+    }
+
+    private void cleanupCoverageFilesOnJenkinsMaster(Run<?, ?> build) {
+        for (File report : getCoverageReports(build)) {
+            report.delete();
         }
+    }
 
-        logger.info(UBERALLS_TAG, "No coverage results found");
-        return null;
+    private Set<File> getCoverageReports(Run<?, ?> build) {
+        Set<File> reports = new HashSet<>();
+
+        File[] foundReports = build.getRootDir().listFiles(COVERAGE_FILENAME_FILTER);
```

Comment:
should we cleanup jacoco and cobertura reports as well?

Reply:
The files are always renamed to `coverage*.xml` during the copy step

Reply:
right, my bad.